Added GCDWebServerAsyncPushResponse response type

KJTsanaktsidis opened the pull request:
Continuing on from the discussion in #118 

In my project, I am implementing a proxy server which takes all requests, replays them to a remote server, and streams the response back to the original requester in my app. This PR adds a kind of buffered responder which you can push data into as you like, providing data to the web server when the appropriate dataReader callbacks are made. Hopefully this should address your concerns about choosing the right time to pass off the data to the  kernel.

I think this would be a good inclusion in the main project because other people doing similar things might find this mode of operation useful, but there is nothing special about this that requires it be in the GCDWebServerTree; I could easily just drop this code into my own project. 

What are your feelings about this kind of response type?

swisspol commented on GCDWebServer/Responses/GCDWebServerAsyncPushResponse.m:
```diff
@@ -0,0 +1,110 @@
+  }
+}
+
+- (void)sendWithData:(NSData *)data {
```
Don't use `@synchronized` it's very heavy handed, use a GCD serial queue instead.

swisspol commented on GCDWebServer/Responses/GCDWebServerAsyncPushResponse.m:
```diff
@@ -0,0 +1,110 @@
+  }
+}
+
+- (void)completeWithErorr:(NSError *)error {
```
Typo

swisspol commented:
I much prefer this approach indeed of putting the impedance logic entirely in the subclass. Does this build without access to the private headers?

KJTsanaktsidis commented:
I have removed the use of @synchronize and fixed the typo. It does build without the private headers, it needs nothing more than the response protocol definition.

KJTsanaktsidis commented:
@swisspol Just wondering if you're still thinking of merging this?

swisspol commented:
I think that based on my comment at https://github.com/swisspol/GCDWebServer/pull/118#issuecomment-68586163, I will respectfully decline to merge this PR. The fact it can be just an extension to GCDWebServer in your code is also a viable workaround.

KJTsanaktsidis commented:
OK, thank you for your help in suggesting this approach and reviewing my code. I will bring these extensions into our own project.

KJTsanaktsidis commented:
In case anybody was interested in what I ended up putting into my project, it was this:
https://gist.github.com/KJTsanaktsidis/12ae946664fc3c1685d2
